syne_tune/optimizer/schedulers/searchers/random_grid_searcher.py

```diff
@@ -136,12 +147,45 @@ def _update(self, trial_id: str, config: dict, result: dict):
             msg = f"Update for trial_id {trial_id}: metric = {metric_val:.3f}"
             logger.info(msg)
 
+    def register_pending(
+        self,
+        trial_id: str,
+        config: Optional[dict] = None,
+        milestone: Optional[int] = None,
+    ):
+        if self._allow_duplicates and trial_id not in self._config_for_trial_id:
+            if config is not None:
+                self._config_for_trial_id[trial_id] = config
+            else:
+                logger.warning(
+                    f"register_pending called for trial_id {trial_id} without passing config"
+                )
+
+    def evaluation_failed(self, trial_id: str):
+        if self._allow_duplicates and trial_id in self._config_for_trial_id:
+            # Blacklist this configuration
+            self._excl_list.add(self._config_for_trial_id[trial_id])
```

Comment:
I've seen these functions now a couple of times. Is it possible to move it up in the class hierarchy or use a utils function instead?

Reply:
Yes, doing that